### PR TITLE
Optimize the code, the logic is clearer

### PR DESCRIPTION
**Pull request by shichanglin5**
## Description

1. putting the sorting directly here makes the code easier to understand2. 
2. use `<=` instead of `<`

## Motivation and Context

Optimize the code, the logic is clearer

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

### CODE REVIEW

File: cmd/endpoint-ellipses.go [hunk 1]
```diff
@@ -66,8 +66,12 @@ var isValidSetSize = func(count uint64) bool {
 }
 
 func commonSetDriveCount(divisibleSize uint64, setCounts []uint64) (setSize uint64) {
+	sort.Slice(setCounts, func(i, j int) bool {
+		return setCounts[i] < setCounts[j]
+	})
+
```
Comment (harshavardhana) [inline cmd/endpoint-ellipses.go]: how is it clean if you remove comments? @shichanglin5

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
| mint-erasure.sh | :x: [more...](#22e0a24a-d991-493a-b084-55633bd63b9f) |
<a name="22e0a24a-d991-493a-b084-55633bd63b9f"></a>
### 15307-4c7d084/mint-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s7:32728
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/14) Running awscli tests ... done in 1 minutes and 2 seconds
(2/14) Running aws-sdk-go tests ... done in 0 seconds
(3/14) Running aws-sdk-java tests ... done in 1 seconds
(4/14) Running aws-sdk-php tests ... done in 43 seconds
(5/14) Running aws-sdk-ruby tests ... done in 5 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 26 seconds
(8/14) Running minio-go tests ... done in 57 seconds
(9/14) Running minio-java tests ... FAILED in 1 minutes and 3 seconds
{
  "name": "minio-java",
  "function": "listObjects()",
  "args": "[bucket, prefix, recursive, 1050 objects]",
  "duration": 60487,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = BucketNotEmpty, message = The bucket you tried to delete is not empty, bucketName = minio-java-test-24ngo3d, objectName = null, resource = /minio-java-test-24ngo3d, requestId = 1701F08E4E80E8AE, hostId = 14eebe61-05d9-479c-a7f9-7f32be28e278)\nrequest={method=DELETE, url=http://minio-k8s7:32728/minio-java-test-24ngo3d, headers=Host: minio-k8s7:32728\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.4.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20220715T073711Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20220715/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=409, headers=Accept-Ranges: bytes\nContent-Length: 328\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO\nStrict-Transport-Security: max-age=31536000; includeSubDomains\nVary: Origin\nVary: Accept-Encoding\nX-Amz-Request-Id: 1701F08E4E80E8AE\nX-Content-Type-Options: nosniff\nX-Xss-Protection: 1; mode=block\nDate: Fri, 15 Jul 2022 07:37:11 GMT\n}\n >>> [io.minio.S3Base$1.onResponse(S3Base.java:690), okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:519), java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128), java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628), java.base/java.lang.Thread.run(Thread.java:829)]"
}
(9/14) Running minio-js tests ... done in 59 seconds
(10/14) Running minio-py tests ... done in 50 seconds
(11/14) Running s3cmd tests ... done in 44 seconds
(12/14) Running s3select tests ... done in 1 seconds
(13/14) Running versioning tests ... done in 3 minutes and 2 seconds

Executed 13 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

File: cmd/endpoint-ellipses.go (hunk 1 above)
Comment (shichanglin5) [reply cmd/endpoint-ellipses.go]: It is easy to understand this code based on this context, I think [Comments rules](https://gist.github.com/wojteklu/73c6914cc446146b8b533c0988cf8d29#comments-rules) is good

If you think the comment is necessary, I can add the comment

Comment (klauspost) [reply cmd/endpoint-ellipses.go]: Yes please.

Comment (shichanglin5) [reply cmd/endpoint-ellipses.go]: okay

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — APPROVED** [review]

**Comment by shichanglin5** [comment]
I think this would be better to understand @harshavardhana 


```go
func commonSetDriveCount(divisibleSize uint64, setCounts []uint64) (setSize uint64) {
	// Not necessarily needed but it ensures to the readers
	// eyes that we prefer a sorted setCount slice for the
	// subsequent function to figure out the right common
	// divisor, it avoids loops.
	sort.Slice(setCounts, func(i, j int) bool {
		return setCounts[i] > setCounts[j]
	})

	// Figure out largest value of total_drives_in_erasure_set which results
	// in least number of total_drives/total_drives_erasure_set ratio.
	for _, cnt := range setCounts {
		if divisibleSize%cnt == 0 {
			return cnt
		}
	}
	return setSize
}
```

File: cmd/endpoint-ellipses.go [hunk 2]
```diff
@@ -66,21 +66,19 @@ var isValidSetSize = func(count uint64) bool {
...
-	prevD := divisibleSize / setCounts[0]
 	for _, cnt := range setCounts {
 		if divisibleSize%cnt == 0 {
-			d := divisibleSize / cnt
-			if d <= prevD {
-				prevD = d
-				setSize = cnt
-			}
+			return cnt
 		}
```
Comment (harshavardhana) [inline cmd/endpoint-ellipses.go:82]: we need to go through the entire list of setCounts() @shichanglin5 - to know the optimal setSize - there may be many erasure ratios possible but we need to choose the best. 
 
Your code will invariably choose something wrong with the sorted setCount - the idea behind commonSetDriveCount() is to get the best possible setCount as well not just the "first" one.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/endpoint-ellipses.go (hunk 2 above)
Comment (shichanglin5) [reply cmd/endpoint-ellipses.go:82]: I noticed the comments:
```
// Figure out largest value of total_drives_in_erasure_set which results
// in least number of total_drives/total_drives_erasure_set ratio
```
After sorting setCounts from large to small, traverse forward, as long as `divisibleSize%cnt==0` is satisfied, then current `cnt` is the `largest value of total_drives_in_erasure_set` as mentioned in the comments

Or suppose the value of `setCounts` is [2, 4, 6, 8], can you give an example under what circumstances the same `divisibleSize` returns different values?

**Comment by klauspost** [comment]
Is there any point to this change?

It seems like an all-risk-no-reward change. This is not critical code, so optimizing doesn't really gain anything.

**Comment by shichanglin5** [comment]
> Is there any point to this change?
> 
> It seems like an all-risk-no-reward change. This is not critical code, so optimizing doesn't really gain anything.

well, when I read the source code of minio, I need to spend more time (maybe others will too) thinking about why the code is written in that way, and there can be a simpler way of writing, so sometimes I wonder if there is anything in it I didn't notice that; submitting a pr is a way to verify this, and if the pr is OK, then also improve the code (readability or whatever)

for risk, this code has enough unit test coverage, so it's fine;

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
| mint-erasure.sh | :x: [more...](#6c4c22a4-b503-40ff-a336-daf732aa5a6b) |
<a name="6c4c22a4-b503-40ff-a336-daf732aa5a6b"></a>
### 15307-2059041/mint-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s2:32757
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/14) Running awscli tests ... done in 53 seconds
(2/14) Running aws-sdk-go tests ... done in 0 seconds
(3/14) Running aws-sdk-java tests ... done in 1 seconds
(4/14) Running aws-sdk-php tests ... done in 43 seconds
(5/14) Running aws-sdk-ruby tests ... done in 6 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 27 seconds
(8/14) Running minio-go tests ... done in 1 minutes and 12 seconds
(9/14) Running minio-java tests ... done in 21 seconds
(10/14) Running minio-js tests ... FAILED in 1 minutes and 28 seconds
{
  "name": "minio-js",
  "function": "listenBucketNotification(bucketName, prefix, suffix, events)",
  "args": "bucketName:minio-js-test-ee0762cb-497d-433f-8eda-254935e5081f, events: s3:ObjectCreated:*",
  "duration": 68284,
  "status": "FAIL",
  "error": "Error: socket hang up at connResetException (internal/errors.js:639:14) at Socket.socketOnEnd (_http_client.js:499:23) at endReadableNT (internal/streams/readable.js:1333:12) at processTicksAndRejections (internal/process/task_queues.js:82:21)"
}
(10/14) Running minio-py tests ... done in 50 seconds
(11/14) Running s3cmd tests ... done in 23 seconds
(12/14) Running s3select tests ... done in 2 seconds
(13/14) Running versioning tests ... done in 3 minutes and 1 seconds

Executed 13 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Pull request closed without merging**